metricbeat/beater/metricbeat.go

```diff
@@ -57,45 +57,35 @@ func New(b *beat.Beat, rawConfig *common.Config) (beat.Beater, error) {
 // within the same Module and MetricSet from collection.
 func (bt *Metricbeat) Run(b *beat.Beat) error {
 
-	bt.client = b.Publisher.Connect()
-
-	// Start each module.
-	var cs []<-chan common.MapStr
-
-	for _, mw := range bt.modules {
-		c := mw.Start(bt.done)
-		cs = append(cs, c)
+	var runners []ModuleRunner
+	for _, m := range bt.modules {
+		r := NewModuleRunner(b.Publisher.Connect, m)
+		r.Start()
+		runners = append(runners, r)
 	}
 
-	// Consume data from all modules and publish it. When the modules stop they
-	// close their output channels. When all the modules' channels are closed
-	// PublishChannels exit.
-	var wg sync.WaitGroup
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		PublishChannels(bt.client, cs...)
-	}()
-
 	if bt.config.ReloadModules.IsEnabled() {
 		logp.Warn("EXPERIMENTAL feature dynamic configuration reloading is enabled.")
-		configReloader := NewConfigReloader(bt.config.ReloadModules, b.Publisher)
 
+		configReloader := NewConfigReloader(bt.config.ReloadModules, b.Publisher)
 		// Start
-		wg.Add(1)
 		go func() {
-			defer wg.Done()
 			configReloader.Run()
 		}()
+		defer configReloader.Stop()
```

Comment:
SGTM. Long term probably all modules should be in the registry and we have only 1 place to stop, but currently this would stop the fixed modules after the first reload.